### DISCUSSION TITLE
npm/npx install from github

### DISCUSSION THREAD

**Discussion by aathan** [NomicFoundation/hardhat#2606] [Troubleshooting]
I'd like to use the very latest patched hardhat and have tried `npm install 'https://github.com/NomicFoundation/hardhat.git#master'` but I end up with a "root" folder in node_modules and `npx hardhat` doesn't think anything got installed.

This is probably more of an npm/npx (and/or yarn?) question but it would save me a lot of time if someone can point out the problem or tell me hardhat can't be installed directly from github.

**Comment by aathan** [comment]
What I did:

```
git clone --depth 1 'https://github.com/NomicFoundation/hardhat.git'
sudo apt-get install node-typescript
cd hardhat
npm i --save-dev @types/node
npm run build
cd /some/other/project
npm link ../hardhat/packages/*
npx hardhat
```

Could someone confirm that's "the right way"?

It turned out the patch I wanted is in branch `test-helpers`. To check that out I had to pull down the full repo, and it appears `test-helpers` is on an unrelated history to `master`.  Could you recommend how to run/install hardhat with specific patch(es) or PR(s) included? (Specifically #2581)

**Comment by alcuadrado** [comment]
That looks good, but this is the recommended way of doing this: https://github.com/NomicFoundation/hardhat/blob/master/CONTRIBUTING.md#developing-locally

Note that we have no intention of supporting npm installs directly from github. Those don't play nice with monorepos nor with typescript.

**Reply by aathan** [reply]
Ok, I was not implying that you should support that, only asked because npm docs implied that's a thing but it didn't work.

Thanks for the doc pointer. I'm having trouble finding documentation on `yarn watch` but the developing-locally docs seem to assume you know how to setup whatever is necessary to develop locally. Probably that's an ok assumption given what most developers would know.

Would it be helpful if I submitted a documentation PR that states how to set up a rudimentary such environment? Does it amount to a git clone of the github repo + whatever I learn about yarn watch when you point me to what it is?

I also asked about the branch/tagging methodology for this project, the thrust of which was knowing how to pull down a particular set of fixes before they're necessarily packaged and released. If you have any git cli magic that you know works to apply all of what's in lets say #2581 without manually looking up the commit hashes and merging them individually that'd be great.

Thanks again.

**Reply by alcuadrado** [reply]
Sure, a PR is welcome!

We don't have any particular branching name nomenclature. We use feature branches for everything. In most cases based off `master`, except for highly risky changes.

**Reply by aathan** [reply]
Ok well, I guess I have a blindspot. Originally I cloned the repo, then I looked at #2581 and saw (here I'm just copy pasting the github page's header area as code):

```
Make eth_getStorageAt spec-compliant #2581
 Open
[fvictorio](https://github.com/fvictorio) wants to merge 3 commits into [test-helpers]
(https://github.com/NomicFoundation/hardhat/tree/test-helpers) from [francovictorio/hh-321]
(https://github.com/NomicFoundation/hardhat/tree/francovictorio/hh-321)
```

at the top of the github page, and was unable to successfully pull the necessary patches without (sorry to repeat) individually applying the commit hashes I found in francovictorio/hh-321.

Would you kindly answer in a non-generalized way that doesn't assume I know lots of things about git and github? I.e., usefully for a noob?  What (again) I'm asking is, here I am looking at #2581.  How do I end up with a local copy of hardhat that has applied the #2581 fixes, canonically speaking? Also, a pointer to what you mean by yarn watch pretty please?

**Reply by alcuadrado** [reply]
I'm sorry, but I can't give you a whole explanation, but I think this is what you are looking for: https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/checking-out-pull-requests-locally

**Reply by aathan** [reply]
Ok, I thought maybe this project's handling of PRs was non-standard, because I couldn't seem to pull down the hh-321 branch successfully. I'll have a look at that link. Thank you.
